### PR TITLE
Integration of Jupyter through docker-compose

### PR DESCRIPTION
**Pull request by Vismayak**
This PR introduces the integration of JupyterHub into our development and production environments. 

**Development Setup:**
1. Create a `.env-dev` file following the format in `.env-example`, paying special attention to the `DOCKER_NETWORK_NAME` and filling in `JUPYTERHUB_CRYPT_KEY`.
2. Start the JupyterHub server using the shell command:
   ```
   ./docker-dev.sh jupyter up
   ```
   To stop the server, use:
   ```
   ./docker-dev.sh jupyter down
   ```
   The JupyterHub can be accessed from the front end via the sidebar.
image

**Production Setup:**
1. Set up a `.env` file as per `.env-example`. Ensure to add `JUPYTERHUB_CRYPT_KEY`.
2. Deploy using:
   ```
   docker-compose -f docker-compose.yml -f docker-compose.jupyter.yml up -d
   ```
   The JupyterHub is accessible only via the local Clowder URL with the postfix `/jupyterhub` (default `http://localhost/jupyterhub/`). It requires logging into the Clowder homepage first; direct access is not allowed. Since we dont have the frontend image that has the updated sidebar, login and then access by manually entering URL
   
image

**Notes:**
- The development setup does not use Traefik and directly connects to `http://localhost:8765`.
- The production setup utilizes Traefik for routing to the notebook server.

Please ensure that your environment files are configured correctly before running the setup commands.

### CODE REVIEW

**Review by longshuicy — COMMENTED** [review]
Could you DM me you env setting? I tried following the example env but somehow i can't get to the port. I can see jupyter container is running though.

image

File: frontend/src/components/Layout.tsx [hunk 1]
```diff
@@ -419,6 +420,23 @@ export default function PersistentDrawerLeft(props) {
 						</ListItemButton>
 					</ListItem>
 				</List>
+				<Divider />
+				{/*TODO: Need to make link dynamic */}
+				<List>
+					<ListItem key={"jupyter"} disablePadding>
+						<ListItemButton
+							component={RouterLink}
+							to="http://localhost:8765/"
```
[the remark this answers is not in the document]
Comment (Vismayak) [reply frontend/src/components/Layout.tsx]: I did add a term to the config but could you let me know if this is the correct way to do it

File: frontend/src/components/Layout.tsx [hunk 2]
```diff
@@ -419,6 +420,23 @@ export default function PersistentDrawerLeft(props) {
...
+							target="_blank"
+							rel="noopener noreferrer"
+						>
+							<ListItemIcon>
+								<MenuBookIcon />
```
[the remark this answers is not in the document]
Comment (Vismayak) [reply frontend/src/components/Layout.tsx:436]: Is there any library you have used for this? I tried using an SVG file but it needs a little bit of resizing to be used? Maybe next version?

**Comment by Vismayak** [comment]
Converting it into draft to use traefik to handle redirects to jupyterlab

**Review by longshuicy — APPROVED** [review]
Tested in both dev and prod mode:
- Redirect to jupyter lab and run notebooks ok
- Shutdown 
- Logout seems to work with one caveat that it also logged out on the frontend. Need to refresh to see that reflect
- Multiple users instantiating different jupyter labs and run notebooks ok

* Note when in production, you need to build the frontend image and point the `docker-compose.yml`. Otherwise it will pull the existing 2.0.0.beta.2 image from docker.io

**Comment by ddey2** [comment]
@Vismayak I still cannot run it on dev environment. I can run it on prod and it works fine there.
Also, I see some frontend 'cannot find module' error, you might want to merge main into this. 

@longshuicy If you have run it on both dev and prod, I think we can merge this. I may be missing some config for dev, network name.

**Comment by longshuicy** [comment]
> @Vismayak I still cannot run it on dev environment. I can run it on prod and it works fine there. Also, I see some frontend 'cannot find module' error, you might want to merge main into this.
> 
> @longshuicy If you have run it on both dev and prod, I think we can merge this. I may be missing some config for dev, network name.

I use the same config but change the network name
`docker network ls` will list the current exist network names. 

If you still can't get it run, maybe let's merge early next week?

File: frontend/webpack.config.dev.js [hunk 3]
```diff
@@ -40,6 +43,7 @@ export default {
 				CLOWDER_REMOTE_HOSTNAME: JSON.stringify(
 					process.env.CLOWDER_REMOTE_HOSTNAME
 				),
+				JUPYTERHUB_URL: JSON.stringify(process.env.JUPYTERHUB_URL),
```
[the remark this answers is not in the document]
Comment (Vismayak) [reply frontend/webpack.config.dev.js:46]: Added, I also added a new shell file for production set up. Updated the readme as well so it is easier to start up the system

**Pull request merged**